src/vsr/replica.zig

```diff
@@ -8761,6 +9088,88 @@ pub fn ReplicaType(
             // responsibility to check.
             self.send_message_to_replica(parameters.replica, reply);
         }
+
+        fn send_commit(self: *Self) void {
+            assert(self.status == .normal);
+            assert(self.primary());
+            assert(self.commit_min == self.commit_max);
+
+            if (self.primary_abdicating) {
+                assert(self.primary_abdicate_timeout.ticking);
+                assert(self.pipeline.queue.prepare_queue.count > 0);
+                assert(self.primary_pipeline_pending() != null);
+
+                log.mark.debug("{}: send_commit: primary abdicating (view={})", .{
+                    self.replica,
+                    self.view,
+                });
+                return;
+            }
+
+            const latest_committed_entry = checksum: {
+                if (self.commit_max == self.superblock.working.vsr_state.checkpoint.commit_min) {
+                    break :checksum self.superblock.working.vsr_state.checkpoint.commit_min_checksum;
+                } else {
+                    break :checksum self.journal.header_with_op(self.commit_max).?.checksum;
+                }
+            };
+
+            self.send_header_to_other_replicas_and_standbys(@bitCast(Header.Commit{
+                .command = .commit,
+                .cluster = self.cluster,
+                .replica = self.replica,
+                .view = self.view,
+                .commit = self.commit_max,
+                .commit_checksum = latest_committed_entry,
+                .timestamp_monotonic = self.clock.monotonic(),
+                .checkpoint_op = self.superblock.working.vsr_state.checkpoint.commit_min,
+                .checkpoint_id = self.superblock.working.checkpoint_id(),
+            }));
+        }
+
+        fn send_request_upgrade_to_self(self: *Self) void {
```

Comment:
Do we guard anywhere that nobody else can send us an upgrade request? I think we can check that in `on_message_from_bus`?

Reply:
We do check that `operation == upgrade` implies `client == 0`, but verifying in `on_message_from_bus` is a good idea!

Reply:
Yeah, I was thinking about an evil client sending a message with `client = 0` "manually"